### DISCUSSION TITLE
adding ormar field type "none"

### DISCUSSION THREAD

**Discussion by avico78** [collerek/ormar#179] [Q&A] [answered]
Maybe it is possible but i'm trying to exclude/suppress field from inherited module .
Let say i have an abstrct ormar class containing the basic structure of class I want to reuse in different ormar modules ,
And in example for some module - i want to suppress specific field .

```python


class CustomerBase(ormar.Model):
    class Meta:
        abstract = True
    customer_name: str = ormar.String(max_length=25,default= lambda : fake.first_name())
    filed_to_remove: str = ormar.String(maxlength=2)


class Customer(CustomerBase,ormar.Model):
    class Meta(BaseMetaCustomerTrg):
        tablename: str = "customer_test"
    
    id: int = ormar.Integer(primary_key=True)
#just an example
    filed_to_remove: none= ormar.None()

```

I know i could redefine it as  pydantic_only=True in Customer module ,
 but I think "none" could make more  sense.

**Comment by collerek** [comment]
Hi, I don't like the idea of None file that has really no use.

Your case can be solved by splitting the base:

```python

class FieldMixin:
    filed_to_remove: str = ormar.String(maxlength=2)

class CustomerBase(ormar.Model):
    class Meta:
        abstract = True
    customer_name: str = ormar.String(max_length=25,default= lambda : fake.first_name())

# you don't have to add ormar.Model (it can even fail mro) as CustomerBase already IS ormar.Model
class Customer(CustomerBase):
    class Meta(BaseMetaCustomerTrg):
        tablename: str = "customer_test"
    
    id: int = ormar.Integer(primary_key=True)
    # this one does not have filed_to_remove: str = ormar.String(maxlength=2)

class CustomerExpanded(FieldMixin, CustomerBase):
    class Meta(BaseMetaCustomerTrg):
        tablename: str = "customer_test"
 
    id: int = ormar.Integer(primary_key=True)
    # now it has also filed_to_remove: str = ormar.String(maxlength=2)
```

**Reply by avico78** [reply]
@collerek 
in case we have "almost" identical class/table CustomerA and CustomerB where 
there are identical --> simple.

if CustomerB has additional field - it easy:

```python 

class CustomerA(ormar.Model):
        class Meta(BaseMetaCustomerA):
    .... many fileds

class CustomerB(CustomerA):
       class Meta(BaseMetaCustomerB):
    addtional_filed : str = ormar.String(max_length=25)

```
but if we CustomerB missing some field - u need to involve a middle/abstract table to apply such change,
I think itinconvenience to handle such way where u have many table with similarity .
so it make more sense to just "suppress" field (as None),

I explain the case:
when you have 3 DBs which 30-40 tables in each one -->  source DB
and u need to copy those 3 DBS to new 3 DBS(target)  where there's slight differences between source and target tables .
it make more sense to have such option.

**Reply by collerek** [reply]
I understand your usecase but None field is kind of ugly.

I think it can be solved in more general (and more useful) way with a custom `copy()` method.

so you could use: something like
```python
class CustomerBase(ormar.Model):
    class Meta:
        abstract = True
    customer_name: str = ormar.String(max_length=25,default= lambda : fake.first_name())
    filed_to_remove: str = ormar.String(maxlength=2)


Customer = CustomerBase.copy(copy_name="Customer", 
                             copy_table_name="customer_test", 
                             exclude_fields={"field_to_remove"},
                             # maybe also new fields?
                             additional_fields={"new_field": ormar.Integer()}
                             # optional metadata and databases params
                             )
```
So you can get exact copy (and i.e only change table name) but also add/exclude fields in a easy way.

If you like the idea you can open another issue for this. ;)

**Reply by avico78** [reply]
> I understand your usecase but None field is kind of ugly.
> 
> I think it can be solved in more general (and more useful) way with a custom `copy()` method.
> 
> so you could use: something like
> 
> ```python
> class CustomerBase(ormar.Model):
>     class Meta:
>         abstract = True
>     customer_name: str = ormar.String(max_length=25,default= lambda : fake.first_name())
>     filed_to_remove: str = ormar.String(maxlength=2)
> 
> 
> Customer = CustomerBase.copy(copy_name="Customer", 
>                              copy_table_name="customer_test", 
>                              exclude_fields={"field_to_remove"},
>                              # maybe also new fields?
>                              additional_fields={"new_field": ormar.Integer()}
>                              # optional metadata and databases params
>                              )
> ```
> 
> So you can get exact copy (and i.e only change table name) but also add/exclude fields in a easy way.
> 
> If you like the idea you can open another issue for this. ;)

Not sure it's ugly as it bringing meaning for suppressing it from the class ,
and i think it with they u are suggesting , it probably may required more configuration ,
as u can have same class name (with different Meta() ) / or class lives in different file (same name).

**Reply by collerek** [reply]
Yeah thinking about it again copy doesn't look like a good idea neither.

By ugly i mean it will confuse mypy and code inspection in ide, it's hard to say just by looking what the none field is doing and you would have to redeclare each field you want to remove.

I think Meta parameter would be much cleaner, something like:

```python
class CustomerBase(ormar.Model):
    class Meta:
        abstract = True
    customer_name: str = ormar.String(max_length=25,default= lambda : fake.first_name())
    field_to_remove: str = ormar.String(maxlength=2)
    other_field = ormar.Integer ()


class Customer(CustomerBase,ormar.Model):
    class Meta(BaseMetaCustomerTrg):
        tablename: str = "customer_test"
        exclude_parent_fields  = {"field_to_remove", "other_field"}
    
    id: int = ormar.Integer(primary_key=True)
```

**Comment by collerek** [comment, marked as answer]
In 0.10.10 I added the exclude_parent_fields that allows you to skip fields by name coming from parents models
